### PR TITLE
[DeviceMesh] Added `DeviceMesh.from_group()`

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124787
* #124780
* #124768
* #124767
* #124741
* #124651

This PR adds a `DeviceMesh.from_group()` static method to convert an existing process group to a device mesh.

Motivation: We need `DeviceMesh.from_group()` to allow FSDP2 to interoperate with distributed libraries that do not use `DeviceMesh` for all parallelisms.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: torch/distributed/device_mesh.py [hunk 1]
```diff
@@ -434,6 +440,23 @@ def get_group(
                     )
                 return dim_groups
 
+        @staticmethod
+        def from_group(group: ProcessGroup, device_type: str) -> "DeviceMesh":
```
Comment (wanchaol) [inline torch/distributed/device_mesh.py:444]: can we add some documentation?

Also I am thinking how we should support multi-dim from group, i.e. HSDP shard, replicate group to pass in to construct a DeviceMesh, I guess the `group` argument would likely need to be `Union[ProcessGroup, List[ProcessGroup]]`

**Review by wanchaol — APPROVED** [review]
This sounds good to me! have one suggestion

File: torch/distributed/device_mesh.py (hunk 1 above)
Comment (awgu) [reply torch/distributed/device_mesh.py:444]: > Also I am thinking how we should support multi-dim from group, i.e. HSDP shard, replicate group to pass in to construct a DeviceMesh, I guess the group argument would likely need to be Union[ProcessGroup, List[ProcessGroup]]

This makes sense to me. Shall we expand the API later when we need it in a separate PR?

Comment (wanchaol) [reply torch/distributed/device_mesh.py:444]: yeah we don't need to do it in this PR as it's BC compatible if later we expand the API

**Comment by awgu** [comment]
cc: @wanchaol I added some documentation and a simple unit test in `test_device_mesh.py` too.

**Pull request closed without merging**